src/odemis/gui/cont/acquisition.py

```diff
@@ -2272,50 +2274,126 @@ def _on_ac_done(self, future):
 
     @call_in_wx_main
     def _on_alignment_done(self, future):
-        self._tab_data.main.is_aligned.value = True
+        """
+        Called when the calibration is finished (either successfully, cancelled or failed).
+        :param future: (ProgressiveFuture) Calibration future object, which can be cancelled.
+        """
+
         self._tab_data.main.is_acquiring.value = False
+        self._future_connector = None  # reset connection to the progress bar
 
-        # Enable button, hide progress bar
-        self._panel.align_gauge_progress.Hide()
-        self._panel.align_spacer_panel.Show()
-        self._panel.align_lbl_gauge.Show()
-        self._panel.btn_align.SetLabel("Alignment...")
-        self._panel.Parent.Layout()
+        try:
+            future.result()  # wait until the calibration is done
+            self.is_aligned.value = True  # allow acquiring ROAs
+            self._reset_calibration_gui("Calibration successful")
+        except CancelledError:
+            self.is_aligned.value = False  # don't enable ROA acquisition
+            self._reset_calibration_gui("Calibration cancelled")  # update label to indicate cancelling
+            return
+        except Exception as ex:
+            self.is_aligned.value = False  # don't enable ROA acquisition
+            logging.exception("Calibration failed with %s.", ex)
+            self._reset_calibration_gui("Calibration failed")
+            return
+
+    def _reset_calibration_gui(self, text=None):
+        """
+        Set back every element in the calibration panel to be ready for the next calibration.
+        :param text (None or str): A (error) message to display instead of the estimated acquisition time.
+        """
+        self.btn_align.Enable(True)  # enable button again
+        self.btn_align.SetLabel("Alignment")  # change button label back to ready for calibration
+        self.gauge_progress.Hide()  # hide progress bar
+        self._panel.align_spacer_panel.Show()  # show space progress bar and progress bar label
+        self.gauge_label.Show()  # show label progress bar
 
-    def _on_align_state(self, aligned):
-        if aligned:
-            self._panel.align_lbl_gauge.SetLabel("System is aligned.")
+        if text is not None:
+            self.gauge_label.SetLabel(text)
         else:
-            # TODO: time estimation
-            self._panel.align_lbl_gauge.SetLabel("~ 2 seconds")
+            duration = align.fastem.estimate_calibration_time(self.calibrations)
+            self.gauge_label.SetLabel(str(duration) + " seconds")
+
+        self._panel.Layout()
+
+    def on_cancel(self, evt):
+        """
+        Cancel the calibration task.
+        Called during calibration when pressing the cancel button.
+        :param evt: (GenButtonEvent) Button triggered.
+        """
+        align_fastem._executor.cancel()  # all the rest will be handled by on_alignment_done()
```

Comment:
Any reason to make a whole function for this? This could be just a single line where the call happens.

Reply:
Not really. I wanted to make the code similar to the acquisition code. There it is also only one line.